File: App/Ask/AskViewModel.swift
import Foundation
import ReadrKit

/// Drives an "ask the book" CONVERSATION: prepares the index, streams each
/// answer, and keeps the transcript so a follow-up can build on what came
/// before.
///
/// It used to hold a single exchange — asking again wiped the previous answer
/// and re-prompted the model from scratch, so "but roads are technically 3D"
/// arrived with nothing to object to.
@MainActor
final class AskViewModel: ObservableObject {

    /// One question and the answer streaming into it.
    struct Exchange: Identifiable, Equatable {
        let id: UUID
        /// Shown immediately, as sent — the reader's message must appear the
        /// moment they send it, not when the answer starts arriving.
        var question: String
        /// What this answer was allowed to see. Kept per exchange because the
        /// panel's toggle can flip between questions, and the footer under
        /// an answer must describe that answer.
        var scope: ReadingScope
        var answerText: String = ""
        var tier: AssembledContext.Tier? = nil
        var citations: [Citation] = []
        /// The stream failed; the panel's error card carries the detail.
        var failed = false
        var isStreaming = true

        /// True once there is nothing to show and nothing coming.
        var isEmpty: Bool { answerText.isEmpty && !isStreaming }
    }

    @Published private(set) var exchanges: [Exchange] = []
    @Published private(set) var isStreaming = false

    /// The mapped, reader-facing failure sentence (an `HTTPError`'s
    /// `errorDescription` when the transport surfaced one). Nil when there is
    /// no error to show.
    @Published var errorMessage: String?
    /// A concrete next step for the reader, shown beneath `errorMessage` (an
    /// `HTTPError`'s `recoverySuggestion`). Nil when the error carries none.
    @Published var errorRecovery: String?

    /// True when there is a configured provider to answer with. Refreshable so
    /// the panel can recover after the reader connects a provider from its own
    /// empty state (A1) without restarting the app.
    @Published private(set) var hasProvider: Bool

    /// True when the active provider can only answer from the book — Apple's
    /// on-device model. The footer then promises exactly that, and no "wider
    /// knowledge" a 3B model does not have.
    @Published private(set) var answersFromBookOnly: Bool

    /// The most recently routed tier — what the grounding caption and the
    /// "Using relevant passages" label describe.
    var tier: AssembledContext.Tier? {
        exchanges.last(where: { $0.tier != nil })?.tier
    }

    /// Re-resolvable provider binding: the panel calls `refresh()` when the
    /// provider settings sheet dismisses, so a key saved from the empty state
    /// takes effect immediately.
    private let makeService: () -> AskService?
    private var service: AskService?
    private let prepare: () async -> Void
    /// Which provider answered, for diagnostics only (#41). A closure because
    /// the active provider can change while the panel is open, same as
    /// `makeService`.
    private let providerName: () -> String
    private let providerAnswersFromBookOnly: () -> Bool
    private let book: Book
    /// The passage the conversation is currently about, or nil for the book
    /// at large. Set per opening (`open`): the conversation outlives the
    /// panel, and the next ✦ Ask on a different passage points it there.
    @Published private(set) var selection: Selection?
    /// How far the reader has read, as of the latest opening. What a scoped
    /// question is held to; nil (a native PDF page) means every question
    /// is about the whole document and the scope choice is hidden.
    @Published private(set) var frontier: ReadingFrontier?
    /// The reader's scope choice — "Whole book" on — kept on the
    /// conversation, with the transcript it describes: a choice made for
    /// one question holds for the next opening rather than snapping back
    /// while whole-book answers sit in the history the model is shown.
    @Published var wholeBook = false
    /// True while the latest opening was a Recap, for the panel's headline.
    @Published private(set) var openedForRecap = false
    /// The question an opening asked to send on the panel's behalf (Recap),
    /// waiting for a provider and for any stream in flight to finish. Sent
    /// once; a later opening replaces it.
    private var pendingQuestion: String?
    /// True once `open` has pointed the conversation somewhere — before
    /// that it has no frontier, and the Ask tab must open it properly.
    private(set) var hasBeenOpened = false
    /// The stream in flight, so leaving the book or starting over cancels
    /// it instead of letting it run to completion for nobody.
    private var streamTask: Task<Void, Never>?
    /// The last question submitted, the scope and the passage it was asked
    /// under, kept so a Retry re-runs exactly it after an error (A5) — not
    /// the passage the conversation has since been pointed at.
    private(set) var lastRequest: (question: String, scope: ReadingScope, selection: Selection?)?

    /// What the next question is allowed to see.
    var scope: ReadingScope {
        if let frontier, !wholeBook { return .upTo(frontier) }
        return .wholeBook
    }

    /// The scope is NOT an init argument: the panel hands it to every
    /// `ask(_:scope:)` from its toggle, so a panel opened scoped can ask the
    /// next question about the whole book without a new view model.
    init(
        makeService: @escaping () -> AskService?,
        prepare: @escaping () async -> Void,
        book: Book,
        selection: Selection?,
        initialQuestion: String?,
        providerName: @escaping () -> String = { "unknown" },
        answersFromBookOnly: @escaping () -> Bool = { false }
    ) {
        self.makeService = makeService
        self.prepare = prepare
        self.providerName = providerName
        self.providerAnswersFromBookOnly = answersFromBookOnly
        self.book = book
        self.selection = selection
        self.pendingQuestion = initialQuestion
        let resolved = makeService()
        self.service = resolved
        self.hasProvider = resolved != nil
        self.answersFromBookOnly = answersFromBookOnly()
    }

    /// Point the conversation at a new opening: the passage (or none), how
    /// far the reader has read, and, for a Recap, the question to send on
    /// the panel's behalf. The transcript stays — one conversation per
    /// book, per session. A stale error from an earlier question does not:
    /// it was about that question, not this opening.
    func open(_ request: AskRequest) {
        selection = request.selection
        frontier = request.scope.frontier
        openedForRecap = request.initialQuestion != nil
        pendingQuestion = request.initialQuestion
        hasBeenOpened = true
        errorMessage = nil
        errorRecovery = nil
        sendPendingIfReady()
    }

    /// Start over: cancel anything in flight, empty the transcript, keep
    /// the passage and frontier the conversation was last pointed at.
    func startOver() {
        cancel()
        exchanges = []
        errorMessage = nil
        errorRecovery = nil
        lastRequest = nil
        pendingQuestion = nil
        openedForRecap = false
    }

    /// Stop the stream in flight, if any. The answer keeps what arrived.
    func cancel() {
        streamTask?.cancel()
        streamTask = nil
    }

    /// Re-resolve the provider binding. Called when the providers sheet
    /// dismisses so a newly saved key flips the panel out of its empty state.
    func refresh() {
        let resolved = makeService()
        service = resolved
        hasProvider = resolved != nil
        answersFromBookOnly = providerAnswersFromBookOnly()
        sendPendingIfReady()
    }

    /// Ask, under the current scope and about the current passage. The
    /// stream runs as the conversation's own task, so it can be cancelled.
    func submit(_ question: String) {
        let scope = self.scope
        let selection = self.selection
        streamTask = Task { await run(question, scope: scope, selection: selection, replacingLastExchange: false) }
    }

    /// Send the question an opening asked for (the Recap), once there is a
    /// provider and no stream in flight. Called from `open`, from
    /// `refresh`, and when a stream ends — so a Recap opened over a
    /// running answer goes out when that answer is done, not never.
    func sendPendingIfReady() {
        guard let question = pendingQuestion, hasProvider, !isStreaming else { return }
        pendingQuestion = nil
        submit(question)
    }

    /// Re-run the last question after an error (A5), under the scope and
    /// about the passage it was asked with. No-op when nothing has been
    /// asked yet.
    func retry() {
        guard let last = lastRequest else { return }
        streamTask = Task {
            await run(last.question, scope: last.scope, selection: last.selection, replacingLastExchange: true)
        }
    }

    // MARK: - Streaming

    private func run(
        _ question: String, scope: ReadingScope, selection: Selection?, replacingLastExchange: Bool
    ) async {
        let trimmed = question.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        // Ignore re-entrant submits while a stream is already in flight.
        guard !isStreaming else { return }
        // Keep the request — question, scope AND passage — so a Retry
        // re-runs exactly it after a failure.
        lastRequest = (trimmed, scope, selection)
        errorMessage = nil
        errorRecovery = nil

        // A Retry re-runs the SAME turn: drop the failed one rather than
        // stacking a second copy of the question in the transcript.
        if replacingLastExchange, let last = exchanges.last, last.answerText.isEmpty {
            exchanges.removeLast()
        }
        let id = UUID()
        exchanges.append(Exchange(id: id, question: trimmed, scope: scope))

        guard self.service != nil else {
            fail(id, "Connect an AI provider in settings to ask questions.", recovery: nil)
            return
        }

        isStreaming = true
        let history = historyBefore(id, scope: scope)
        defer {
            isStreaming = false
            update(id) { $0.isStreaming = false }
            // A Recap that arrived while this answer streamed goes now.
            sendPendingIfReady()
        }

        await prepare()
        // `prepare` renews expired OAuth tokens (see AskPanelView); providers
        // capture credentials by value, so re-resolve the service to bake the
        // refreshed tokens into the one that streams.
        refresh()
        guard let service else {
            fail(id, "Connect an AI provider in settings to ask questions.", recovery: nil)
            return
        }
        do {
            for try await event in service.ask(
                trimmed, about: book, selection: selection, history: history,
                scope: scope
            ) {
                // Cancelled (the reader left the book or started over):
                // keep what arrived, say nothing.
                if Task.isCancelled { return }
                switch event {
                case let .contextAssembled(tier):
                    update(id) { $0.tier = tier }
                case let .citations(list):
                    update(id) { $0.citations = list }
                case let .token(delta):
                    update(id) { $0.answerText += delta }
                case let .completed(fullText):
                    // Authoritative final text — covers providers that don't
                    // stream incremental deltas.
                    update(id) { $0.answerText = fullText }
                }
            }
            // A small model can end with nothing to show — every sentence it
            // produced was a copy of the passages, or a loop cut at its first
            // sentence. The panel says so in a line; the log says which
            // provider, for the bug report.
            if exchanges.first(where: { $0.id == id })?.answerText
                .trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == true {
                DiagnosticsLog.shared.record(
                    .warning, .provider, "ask: answer came back empty (provider \(providerName()))"
                )
            }
        } catch is CancellationError {
            return
        } catch {
            // Surface the mapped, actionable sentence from the error (A5):
            // `HTTPError` conforms to `LocalizedError` so timeouts, rejected
            // keys, and rate limits read as something the reader can act on
            // instead of Foundation's generic "operation couldn't be
            // completed." The recovery suggestion, when present, is shown
            // beneath it.
            if let localized = error as? LocalizedError {
                fail(
                    id,
                    localized.errorDescription ?? error.localizedDescription,
                    recovery: localized.recoverySuggestion
                )
            } else {
                fail(id, error.localizedDescription, recovery: nil)
            }
            // The provider and routing tier, never the question — that's the
            // reader's, and a bug report is public (#41).
            DiagnosticsLog.shared.recordAskFailure(
                provider: providerName(),
                tier: (exchanges.first { $0.id == id }?.tier)?.rawValue ?? "unrouted",
                error: error
            )
        }
    }

    /// Completed turns before `id`, oldest first — what the model is shown of
    /// the conversation so far. A failed or empty turn carries no answer and
    /// is left out. So is a whole-book turn when THIS question is scoped:
    /// the no-spoilers promise covers the history the model reads, not
    /// just the passages it is handed.
    private func historyBefore(_ id: UUID, scope: ReadingScope) -> [ConversationTurn] {
        exchanges.prefix(while: { $0.id != id }).compactMap { exchange -> ConversationTurn? in
            guard !exchange.failed, !exchange.answerText.isEmpty else { return nil }
            if scope.isScoped, !exchange.scope.isScoped { return nil }
            return ConversationTurn(
                question: exchange.question,
                answer: Answer(
                    text: exchange.answerText,
                    tier: exchange.tier ?? .retrieval,
                    citations: exchange.citations
                )
            )
        }
    }

    private func update(_ id: UUID, _ change: (inout Exchange) -> Void) {
        guard let index = exchanges.firstIndex(where: { $0.id == id }) else { return }
        change(&exchanges[index])
    }

    private func fail(_ id: UUID, _ message: String, recovery: String?) {
        errorMessage = message
        errorRecovery = recovery
        update(id) {
            $0.failed = true
            $0.isStreaming = false
        }
    }
}

File: android/app/src/main/java/com/readrai/readr/ui/ask/AskViewModel.kt
package com.readrai.readr.ui.ask

import android.util.Log
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.compose.viewModel
import com.readrai.readr.ReadrApplication
import com.readrai.readr.data.AskCitation
import com.readrai.readr.data.AskEvent
import com.readrai.readr.data.AskFrontier
import com.readrai.readr.data.AskPosition
import com.readrai.readr.data.AskRepository
import com.readrai.readr.data.AskSelection
import com.readrai.readr.data.AskTier
import com.readrai.readr.data.AskTurn
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * The book's Ask conversation, as a view model bound to this composition —
 * keyed by book, so two books never share a transcript, and scoped to the
 * reader's back-stack entry, so a trip to the provider settings and back
 * finds the same one.
 */
@Composable
fun rememberAskViewModel(bookId: String): AskViewModel {
    val app = LocalContext.current.applicationContext as ReadrApplication
    return viewModel(key = "ask/$bookId") {
        AskViewModel({ app.asks() }, app.askConversations.forBook(bookId))
    }
}

/**
 * The passages one answer leaned on, wrapped so the exchange around them can
 * be `@Immutable`: a bare `List` is not a stable type to Compose, and an
 * exchange holding one is re-composed on every state write in the sheet
 * whether or not anything about it changed.
 */
@Immutable
data class AskCitations(val items: List<AskCitation> = emptyList()) {
    val isEmpty: Boolean get() = items.isEmpty()
}

/**
 * One answer's structure, as the kit's Markdown parser cut it — wrapped for
 * the same reason [AskCitations] is: a bare `List` is not a stable type to
 * Compose, and an exchange holding one is re-composed on every state write in
 * the sheet.
 *
 * Empty while the answer is still streaming. The parse is a bridge call, so
 * it happens ONCE, off the main thread, when the answer is finished — until
 * then the sheet draws the text as paragraphs, which is all a half-written
 * answer has to show anyway.
 */
@Immutable
data class AnswerBlocks(val items: List<AnswerBlock> = emptyList()) {
    val isEmpty: Boolean get() = items.isEmpty()
}

/** One question and the answer streaming into it. */
@Immutable
data class AskExchange(
    val id: Long,
    /** Shown as sent the moment it is sent, not when the answer starts. */
    val question: String,
    /**
     * Whether this answer was held to what the reader had read. Kept per
     * exchange because the sheet's scope control can flip between questions,
     * and the note under an answer must describe that answer.
     */
    val scoped: Boolean,
    val answerText: String = "",
    val tier: String? = null,
    /** What the routed tier promises, in the kit's own words. */
    val providesCitations: Boolean? = null,
    val citations: AskCitations = AskCitations(),
    /**
     * The finished answer's blocks. Empty while it streams — the sheet draws
     * paragraphs then — and filled once, off the main thread, when it lands.
     */
    val blocks: AnswerBlocks = AnswerBlocks(),
    /**
     * Why this turn has no answer — the kit's sentence, kept ON the exchange
     * so it stays under the question it belongs to. The composer's error card
     * can be cleared by the next question; the transcript must still say what
     * happened to this one.
     */
    val failure: String? = null,
    val failureRecovery: String? = null,
    val isStreaming: Boolean = true,
) {
    val failed: Boolean get() = failure != null

    /** Nothing to show and nothing coming. */
    val isEmpty: Boolean get() = answerText.isBlank() && !isStreaming && !failed
}

/**
 * A book's Ask conversation, for the life of the process — the Apple app
 * keeps one per session the same way. It outlives the sheet and the reader's
 * view model, so closing Ask and opening it again on the next chapter
 * continues the conversation rather than starting a new one.
 */
class AskConversation(val bookId: String) {
    var exchanges by mutableStateOf<List<AskExchange>>(emptyList())
    /**
     * The reader's scope choice, kept with the transcript it describes: a
     * choice made for one question holds for the next opening rather than
     * snapping back while whole-book answers sit in the history the model
     * reads.
     */
    var wholeBook by mutableStateOf(false)

    private var nextId = 1L

    fun nextExchangeId(): Long = nextId++

    /** Start over, keeping the same object so open sheets stay bound to it. */
    fun clear() {
        exchanges = emptyList()
    }
}

/** The process's conversations, one per book. */
class AskConversations {
    private val byBook = HashMap<String, AskConversation>()

    @Synchronized
    fun forBook(bookId: String): AskConversation = byBook.getOrPut(bookId) { AskConversation(bookId) }

    /**
     * Drop a book's transcript. A removed book takes its conversation with
     * it: the answers quote a book nobody can open any more, and a re-import
     * is a new book with a new id in any case.
     */
    @Synchronized
    fun forget(bookId: String) {
        byBook.remove(bookId)
    }
}

/**
 * Everything one opening of the sheet needs, decided by whoever opens it: the
 * ✦ in the reader bar opens on the book at the reader's place, the capsule's
 * ✦ Ask opens on the passage.
 */
data class AskRequest(
    /** The passage the question is about, or null for a book-wide question. */
    val selection: AskSelection? = null,
    /**
     * How far the reader has read. Null means there is no reading position to
     * scope to, so every question is about the whole book and the scope
     * control is not offered.
     */
    val frontier: AskFrontier? = null,
)

/**
 * Drives one book's Ask conversation: streams each answer, keeps the
 * transcript so a follow-up can build on what came before, and remembers the
 * last request so a failure can be retried without retyping.
 *
 * Nothing here writes a reader-facing failure sentence: the facade hands over
 * ReadrKit's own message and recovery line, which is what the error card
 * shows.
 */
class AskViewModel(
    private val openRepository: suspend () -> AskRepository,
    private val conversation: AskConversation,
) : ViewModel() {

    val bookId: String get() = conversation.bookId
    val exchanges: List<AskExchange> get() = conversation.exchanges

    /** "Whole book" chosen over "Up to where I am". */
    var wholeBook: Boolean
        get() = conversation.wholeBook
        set(value) { conversation.wholeBook = value }

    var isStreaming by mutableStateOf(false)
        private set
    /** The book's retrieval index is being built; the answer waits on it. */
    var indexing by mutableStateOf(false)
        private set

    /** The kit's sentence for the failure, and the step it suggests. */
    var errorMessage by mutableStateOf<String?>(null)
        private set
    var errorRecovery by mutableStateOf<String?>(null)
        private set

    /** Null while the answer is still being fetched from the kit. */
    var hasProvider by mutableStateOf<Boolean?>(null)
        private set

    /**
     * True when the active model runs on the phone: it answers from the
     * passages and nothing else, and the grounding caption says exactly that
     * rather than promising a wider knowledge it does not have.
     */
    var answersFromBookOnly by mutableStateOf(false)
        private set

    /**
     * What to do about having nothing connected. The facade's sentence names
     * only the doors THIS phone has — a key, and the phone's own model where
     * it can run one — but the empty state must never be a heading over a
     * blank line, so it starts on the one door every build has and is
     * replaced only by a sentence that actually says something.
     */
    var setupGuidance by mutableStateOf(DEFAULT_SETUP_GUIDANCE)
        private set

    /** The passage this opening pointed the conversation at, or null. */
    var selection by mutableStateOf<AskSelection?>(null)
        private set
    var frontier by mutableStateOf<AskFrontier?>(null)
        private set
    /** "Chapter 7 of 24 · 31% · The Whale" — the kit's own line. */
    var position by mutableStateOf<AskPosition?>(null)
        private set

    /**
     * Whether the sheet should be showing. Kept here rather than in the
     * reader's composition so a trip to the provider settings and back comes
     * home to the sheet the reader left open.
     */
    var isOpen by mutableStateOf(false)
        private set

    /** The question, scope and passage to re-run after a failure. */
    var lastRequest: AskRequest? = null
        private set
    private var lastQuestion: String? = null

    private var repository: AskRepository? = null
    private var streamJob: Job? = null

    /** What the next question is allowed to see. */
    val scopedFrontier: AskFrontier? get() = frontier?.takeIf { !wholeBook }
    val isScoped: Boolean get() = scopedFrontier != null

    /** The most recently routed tier — what the tier label describes. */
    val tier: String? get() = exchanges.lastOrNull { it.tier != null }?.tier

    /**
     * What the last routed tier promises about citations, in the kit's own
     * answer — null until something has actually been routed, which is what
     * keeps the caption from promising anything before then.
     */
    val providesCitations: Boolean?
        get() = exchanges.lastOrNull { it.providesCitations != null }?.providesCitations

    /**
     * Point the conversation at a new opening. The transcript stays; a stale
     * error does not — it was about the last question, not this opening.
     */
    fun open(request: AskRequest) {
        selection = request.selection
        frontier = request.frontier
        errorMessage = null
        errorRecovery = null
        position = null
        isOpen = true
        refresh()
        loadPosition()
        parseMissingBlocks()
    }

    /**
     * The sheet went away. A stream in flight is left running — the answer
     * belongs to the conversation, not to the sheet, and it will be there
     * when the reader comes back.
     */
    fun close() {
        isOpen = false
    }

    /** Re-resolve the provider, and the sentence the empty state shows. */
    fun refresh() {
        viewModelScope.launch {
            val repo = repository() ?: return@launch
            hasProvider = runCatching { repo.hasProvider() }.getOrDefault(false)
            answersFromBookOnly = runCatching { repo.answersFromBookOnly() }.getOrDefault(false)
            runCatching { repo.setupGuidance("ask questions") }
                .getOrNull()
                ?.takeIf { it.isNotBlank() }
                ?.let { setupGuidance = it }
        }
    }

    /** Start over: cancel anything in flight and empty the transcript. */
    fun startOver() {
        cancel()
        conversation.clear()
        errorMessage = null
        errorRecovery = null
        lastRequest = null
        lastQuestion = null
    }

    /** Stop the stream in flight. The answer keeps whatever arrived. */
    fun cancel() {
        streamJob?.cancel()
        streamJob = null
        isStreaming = false
        indexing = false
        // What arrived before the stop is the answer now; the stream's own
        // `finally` gets the last word on it, including its blocks.
        update(exchanges.lastOrNull()?.id) { it.copy(isStreaming = false) }
    }

    /** Ask, under the current scope and about the current passage. */
    fun submit(question: String) {
        val request = AskRequest(selection = selection, frontier = scopedFrontier)
        streamJob = viewModelScope.launch { run(question, request, replacingLast = false) }
    }

    /** Re-run the last question exactly as it was asked. */
    fun retry() {
        val request = lastRequest ?: return
        val question = lastQuestion ?: return
        streamJob = viewModelScope.launch { run(question, request, replacingLast = true) }
    }

    /** Static starters, worded for the scope and the passage. */
    val suggestions: List<String>
        get() = when {
            selection != null -> listOf(
                "What does this passage mean?",
                if (isScoped) "How does this connect to what I've read so far?"
                else "How does this connect to the rest of the book?",
            )
            isScoped -> listOf(
                RECAP_QUESTION,
                "Summarize what I've read so far",
                "What are the key themes so far?",
                "Who are the main characters so far?",
            )
            else -> listOf(
                "Summarize this book",
                "What are the key themes?",
                "Who are the main characters?",
            )
        }

    // MARK: - Streaming

    private suspend fun run(question: String, request: AskRequest, replacingLast: Boolean) {
        val trimmed = question.trim()
        if (trimmed.isEmpty() || isStreaming) return
        lastRequest = request
        lastQuestion = trimmed
        errorMessage = null
        errorRecovery = null

        // A retry re-runs the SAME turn: drop the failed one rather than
        // stacking a second copy of the question in the transcript.
        if (replacingLast && conversation.exchanges.lastOrNull()?.answerText.isNullOrBlank()) {
            conversation.exchanges = conversation.exchanges.dropLast(1)
        }
        val id = conversation.nextExchangeId()
        val scoped = request.frontier != null
        conversation.exchanges = conversation.exchanges + AskExchange(id = id, question = trimmed, scoped = scoped)

        val repo = repository()
        if (repo == null) {
            fail(id, COULD_NOT_OPEN, null)
            return
        }
        isStreaming = true
        val history = historyBefore(id)
        // Deltas arrive a few characters at a time and each one would
        // otherwise be a state write, a re-parse of the answer and a re-layout
        // of the sheet. They are joined here and written at most this often,
        // which is still faster than anyone reads.
        val pending = StringBuilder()
        var lastWrite = 0L
        fun flush() {
            if (pending.isEmpty()) return
            val delta = pending.toString()
            pending.setLength(0)
            update(id) { it.copy(answerText = it.answerText + delta) }
        }
        try {
            repo.ask(bookId, trimmed, request.frontier, request.selection, history).collect { event ->
                when (event) {
                    AskEvent.Indexing -> indexing = true
                    is AskEvent.Routed -> {
                        indexing = false
                        update(id) { it.copy(tier = event.tier, providesCitations = event.providesCitations) }
                    }
                    is AskEvent.Citations -> update(id) { it.copy(citations = AskCitations(event.citations)) }
                    is AskEvent.Token -> {
                        pending.append(event.text)
                        val now = System.currentTimeMillis()
                        if (now - lastWrite >= COALESCE_MILLIS) {
                            lastWrite = now
                            flush()
                        }
                    }
                    // Authoritative final text — it covers providers that do
                    // not stream incremental deltas.
                    is AskEvent.Completed -> {
                        pending.setLength(0)
                        update(id) { it.copy(answerText = event.text) }
                    }
                    is AskEvent.Failed -> {
                        flush()
                        fail(id, event.message, event.recovery)
                    }
                }
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            // The class, never the message: an exception on this path can
            // carry a URL or a payload, and a log is not a place for either.
            Log.w(TAG, "ask failed: ${e.javaClass.simpleName}")
            fail(id, COULD_NOT_OPEN, null)
        } finally {
            flush()
            isStreaming = false
            indexing = false
            update(id) { it.copy(isStreaming = false) }
            // The answer is whatever arrived — completed, failed part-way, or
            // stopped by the reader. Whatever it is, it is finished, so it is
            // parsed now rather than by the composition drawing it.
            parseBlocks(id)
        }
    }

    /**
     * The answered turns before `id`, oldest first. A failed or empty turn
     * carries no answer and is left out; each surviving turn says whether it
     * was answered under a scope, and the facade drops the unscoped ones from
     * a scoped question's history.
     */
    private fun historyBefore(id: Long): List<AskTurn> =
        conversation.exchanges
            .takeWhile { it.id != id }
            .filter { !it.failed && it.answerText.isNotBlank() }
            .map {
                AskTurn(
                    question = it.question,
                    answerText = it.answerText,
                    tier = it.tier ?: AskTier.RETRIEVAL,
                    citations = it.citations.items,
                    scoped = it.scoped,
                )
            }

    private suspend fun repository(): AskRepository? {
        repository?.let { return it }
        return try {
            openRepository().also { repository = it }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "ask repository failed: ${e.javaClass.simpleName}")
            hasProvider = false
            null
        }
    }

    private fun loadPosition() {
        val place = frontier ?: return
        viewModelScope.launch {
            val repo = repository() ?: return@launch
            position = runCatching { repo.positionSummary(bookId, place.chapterIndex, place.utf16Offset) }.getOrNull()
        }
    }

    /**
     * Cut a finished answer into the kit's blocks and hang them on the
     * exchange it belongs to.
     *
     * Once, and off the main thread: the parse is a JNI call and a JSON
     * decode, and the sheet used to make it from composition — on every
     * delta, on the thread that then had to lay the result out. Composition
     * now reads what is already on the exchange and never crosses the bridge
     * at all. The text is checked again on the way back, so an answer that
     * moved on meanwhile keeps its own blocks.
     */
    private fun parseBlocks(id: Long) {
        val text = conversation.exchanges.firstOrNull { it.id == id }?.answerText ?: return
        if (text.isBlank()) return
        viewModelScope.launch {
            val repo = repository() ?: return@launch
            // `Default`, not `IO`: the bridge call is a parse — CPU, no disk
            // and no socket — and it must not sit behind a provider request.
            val parsed = runCatching {
                withContext(Dispatchers.Default) {
                    AnswerMarkdown.blocks(repo.answerBlocksJSON(text), fallback = text)
                }
            }.getOrElse { listOf(AnswerBlock.Paragraph(text)) }
            update(id) { if (it.answerText == text) it.copy(blocks = AnswerBlocks(parsed)) else it }
        }
    }

    /**
     * Blocks for answers that are already in the transcript without them — a
     * conversation restored on a later opening of the sheet, and an answer
     * whose stream was cancelled rather than completed.
     */
    private fun parseMissingBlocks() {
        for (exchange in conversation.exchanges) {
            if (exchange.blocks.isEmpty && !exchange.isStreaming && exchange.answerText.isNotBlank()) {
                parseBlocks(exchange.id)
            }
        }
    }

    private fun update(id: Long?, change: (AskExchange) -> AskExchange) {
        if (id == null) return
        conversation.exchanges = conversation.exchanges.map { if (it.id == id) change(it) else it }
    }

    private fun fail(id: Long, message: String, recovery: String?) {
        errorMessage = message
        errorRecovery = recovery
        update(id) { it.copy(failure = message, failureRecovery = recovery, isStreaming = false) }
    }

    companion object {
        /**
         * The recap, word for word — the first suggestion chip when answers
         * are scoped, and the same sentence the Apple reader sends.
         */
        const val RECAP_QUESTION = "Recap what I've read so far — no spoilers"

        /** The only sentence this class writes, and only when the kit is unreachable. */
        private const val COULD_NOT_OPEN = "Readr couldn't reach the model. Try asking again."
        private const val TAG = "Readr.Ask"

        /** One state write per frame or two, however fast the tokens come. */
        private const val COALESCE_MILLIS = 50L

        /**
         * The empty state's opening sentence, before the facade has said what
         * this phone offers. Every build of this app has the key path, so it
         * is true of all of them; the facade's own sentence replaces it as
         * soon as it arrives.
         */
        const val DEFAULT_SETUP_GUIDANCE = "Add an API key to ask questions."
    }
}
